### DISCUSSION TITLE
Struggling to upgrade to v0.48

### DISCUSSION THREAD

**Discussion by bhagdave** [libp2p/rust-libp2p#3198] [Q&A] [answered]
Hi All,

Not sure if this is the right place to ask this question but here goes.

I wanted to start playing with libp2p and started with the example at [Log Rocket](https://blog.logrocket.com/libp2p-tutorial-build-a-peer-to-peer-app-in-rust/) but found the version of libp2p being used was 0.31 so took a copy of the [code repo](https://github.com/zupzup/rust-peer-to-peer-example) and started to make the required changes to move it to version 0.47.  I now have that working [here](https://github.com/bhagdave/p2p-play) but when I change the libp2p version to 0.48 I hit a problem with one of the elements of the Behaviour struct (which had `#[behaviour(ignore)]`) not having a From implementation.  

The error I am receiving is:-
```
error[E0277]: the trait bound `StoryBehaviourEvent: From<<tokio::sync::mpsc::UnboundedSender<ListResponse> as NetworkBehaviour>::OutEvent>` is not satisfied
  --> src/main.rs:64:10
   |
64 | #[derive(NetworkBehaviour)]
   |          ^^^^^^^^^^^^^^^^ the trait `From<<tokio::sync::mpsc::UnboundedSender<ListResponse> as NetworkBehaviour>::OutEvent>` is not implemented for `StoryBehaviourEvent`
   |
   = help: the following other types implement trait `From<T>`:
             <StoryBehaviourEvent as From<FloodsubEvent>>
             <StoryBehaviourEvent as From<MdnsEvent>>
   = help: see issue #48214
   = note: this error originates in the derive macro `NetworkBehaviour` (in Nightly builds, run with -Z macro-backtrace for more info)

```
I have searched through issues and discussions and the nearest issue I can find is the problems related to using Generics in the struct and I don't think that is relevant here.

Can anyone point me in the right direction?  Thanks in advance.

**Comment by thomaseizinger** [comment, marked as answer]
> Not sure if this is the right place to ask this question 

It is!

`0.31` is quite an old version and a few things have changed. It might be a bit much but you could read through the [CHANGELOG](https://github.com/libp2p/rust-libp2p/blob/master/CHANGELOG.md) to figure out how to migrate that code.

You are probably better off by looking at one of the existing examples: https://github.com/libp2p/rust-libp2p/tree/master/examples

Does the compile output any deprecation warnings at all on version 0.47? We usually try to guide users through breaking changes with that. `NetworkBehaviourEventProcess` and the ability to ignore fields has been removed in `libp2p-swarm-derive v0.30.0`: https://github.com/libp2p/rust-libp2p/blob/master/swarm-derive/CHANGELOG.md#0300 in favor of tracking that data wherever you are polling `Swarm`!

**Reply by bhagdave** [reply]
I had gone through the changelog to get it up to 0.47 and had made the required changes to make sure I had no deprecation warnings when compiling on 0.47.  Which is why I think I got a bit stumped when I got the errors when moving to 0.48.  

I did look through the examples but could not find one that matched my scenario.

Anyway I will check the swarm derive change log as suggested and see where I get.  

THANKS
@thomaseizinger Confimed as Answer.  I moved the element out of my NetworkBahviour struct and it compiles fine on 0.48

**Reply by thomaseizinger** [reply]
Glad it is sorted!

The examples should demonstrate how to get information like mDNS stuff from the event into another behaviour. Basically you match on the return value of `swarm.next()` and pass it on to the relevant sub-behaviour again.
